src/server.rs

```diff
@@ -0,0 +1,161 @@
+pub struct Stats {
+    failure: usize,
+    success: usize,
+}
+
+impl Stats {
+    pub fn new() -> Stats {
+        Stats {
+            failure: 0,
+            success: 0,
+        }
+    }
+
+    pub fn inc_success(&mut self) {
+        self.success += 1;
+    }
+
+    pub fn inc_failure(&mut self) {
+        self.failure += 1;
+    }
+
+    pub fn success(&self) -> usize {
+        self.success
+    }
+
+    pub fn failure(&self) -> usize {
+        self.failure
+    }
+}
+
+/// A wrapper around the internals of Server
+///
+/// All functions hide the RwLock away so that we can ensure the lock is not held for a long time.
+#[derive(Clone, Debug)]
+pub struct Server {
+    inner: Arc<RwLock<inner::Server>>,
+}
+
+impl Server {
+    pub fn new(url: Url) -> Server {
+        let inner = inner::Server::new(url);
+
+        Server {
+            inner: Arc::new(RwLock::new(inner)),
+        }
+    }
+
+    pub fn with_map_host(self, map_host: bool) -> Self {
+        let rwlock = Arc::try_unwrap(self.inner).expect("Multiple references");
+        let server = rwlock.into_inner().expect("Lock is poisoned");
+        let inner = server.with_map_host(map_host);
+
+        Server {
+            inner: Arc::new(RwLock::new(inner)),
+        }
+    }
+
+    pub fn url(&self) -> Url {
+        self.inner.read().expect("Lock is poisoned").url()
+    }
+
+    pub fn map_host(&self) -> bool {
+        self.inner.read().expect("Lock is poisoned").map_host()
+    }
+
+    pub fn inc_success(&self) {
+        self.inner.write().expect("Lock is poisoned").stats_mut().inc_success();
+    }
+
+    pub fn inc_failure(&self) {
+        self.inner.write().expect("Lock is poisoned").stats_mut().inc_failure();
+    }
+
+    pub fn stats(&self) -> Stats {
+        self.inner.read().expect("Lock is poisoned").stats().clone()
+    }
+}
+
+impl FromStr for Server {
+    type Err = ::hyper::error::ParseError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let url: Url = if s.starts_with("http") {
+            try!(s.parse())
+        } else {
+            try!(format!("http://{}", s).parse())
+        };
+        Ok(Server::new(url))
+    }
+}
+
+/// Check if two servers are likely the same
+///
+/// This check is used by the Pool to find another server. The comparison is shallow as we are not
+/// looking for a server with the exact same stats.
+impl PartialEq for Server {
+    fn eq(&self, other: &Server) -> bool {
+        self.url() == other.url()
+    }
+}
+
+mod inner {
+    use hyper::Url;
+
+    use super::Stats;
+
+    #[derive(Clone, Debug, Eq, PartialEq)]
+    pub struct Server {
+        url: Url,
+        /// Whether to use the server host name in the Host header when making a backend request
+        ///
+        /// The default case is for the upstream Host header to be used. Some origins, such as
+        /// Heroku or Amazon S3, must use the backend server hostname.
+        map_host: bool,
+        hc_failure: usize,
```

Comment:
It is a good question. I imagine we could have a separate statistics module that keys off of something like server name. 

I think we must have `hc_failure` on the `Server` struct in order to know when to evict it. Whether we do passive or active health checks, we need to track it local to that backend. Also, the server will probably have some sort of _state_ enum attached to it at some point. Think `active`, `down`, `disabled`, etc.

Do you want to re-think this before merging in? I tried 2 other strategies to not require `Arc` and locks, but could not get those to work.